### PR TITLE
Add ability to load BD/BBD plugins

### PR DESCRIPTION
**Pull request by rauenzi**
**Note:** I suggest that this is tested by several people before fully merging. I have personally tested several plugins from many plugin devs of BD and all that I have tested are working seamlessly. But it never hurts to have additional testing.

This PR adds the ability to load BD plugins as normal plugins in ED. This also creates a decent API for plugins (`EDApi`) that is also used for the BD API (`BdApi`). Since BD plugins already require it, I figured I would add an ED-specific alias so ED-specific plugins may also make use of it. 

Some things of note include:
1. BD plugin settings will not appear in the setting tabs because they don't have normalized settings schemes
2. BD plugins that have settings will have a settings button next to them in the plugin list that will open a modal with the plugin's settings
3. Some seemingly arbitrary variables/objects are added to the global scope, this is due to BD's previously horrible architecture and BD plugin tendencies to take advantage of anything BD puts on the global scope so many of these are unfortunately required
4. jQuery is added when the `BDManager` is setup, this is due to *many* BD plugins making use of it
5. An observer is setup internally in the `BDManager` to pass to plugin's `observer` function, something that is again required through BD and several plugins make use of

File-by-file changes are found below:

dom_shit.js
1. Added the EDApi/BdApi with several functions required by BD plugins
2. New require of the BDManager and the setup before loading plugins

plugins/ed_settings.js
1. Add settings button for BD plugins
2. Adjust click listener to distinguish between buttons

+bd_shit.js
1. Add override of  `require` to allow BD plugins to be loaded through it
2. Add ability to convert/wrap a BD plugin to a compatible ED plugin
3. Create function to add unfortunately necessary global scope items
4. Add observer and did-navigate listener for required BD functions `observer` and `onSwitch`
5. Add jQuery dependency
6. Add minimal CSS (normally added by BD) that almost all BD plugins depend on for their settings to not look like crap
7. Add function to create and show the settings modal for BD plugins

### CODE REVIEW

**Comment by casKd-dev** [comment]
I am looking forward for plugin support from BD, or a conventional way to convert them. +1 from me. Great Job!

**Comment by rauenzi** [comment]
@casKd-dev Thanks you can test it yourself by adding `bd_shit.js` to your ED folder, replacing the `dom_shit.js` with the one from this PR and replacing the `plugins/ed_settings.js` with the one from this PR

**Comment by JiMiKi** [comment]
@rauenzi I followed your instructions to casKd-dev to enable BD plugins in ED. 
Everything seemed fine at first, BD Plugin support showed up in ED Settings and BD Plugins appeared in the Plugins list, however none of the BD plugins would actually work, at all. 
So me and a friend tried to figure out how to make it work, and my friend figured out how to make them work. 
We had to remove the file called "pluginapi" in our ED plugin folder. 

So TL;DR - for the steps mentioned above to work, we had to also remove pluginapi.js from ED's plugin folder.

Wasn't sure if you were aware of this or not, so figured I'd post this here.

**Comment by rauenzi** [comment]
This is mentioned in the FAQ of the server. You *cannot* mix my ED and BD plugins, you have one set or the other as the libraries needed for either one conflict.

**Comment by JiMiKi** [comment]
@rauenzi I see, I was not aware of this, and I'm afraid I can't find where this FAQ is, I am in Zere's Place and EnhancedDiscord servers. 
I might be blind but I can't find a FAQ or a trace of that information anywhere, so if you could guide me to it, that would be greatly appreciated! 
Also, thanks for making this bridge between BD and ED, really missed a lot of plugins from BD!

**Comment by rauenzi** [comment]
There is literally a channel named "faq" in the ED server

**Comment by JiMiKi** [comment]
Oh, there is... well... as I said, I'm blind... Thanks for the help and sorry for the ignorance!

**Comment by rauenzi** [comment]
This has been updated to be merge-able and also includes the patch from #54

**Comment by rauenzi** [comment]
>Do I just drop 0PluginLibrary.plugin.js in plugins folder for those plugins that need it?

Not really relevant to the PR but yes

>And if so, should I be concerned about this error?

Again, this is specific to that plugin, but no it's non-fatal.

**Comment by jakuski** [comment]
Quick little idea for the shared `EDAp / BdApi`:

Since you are already "linking" React and React-DOM in it, why don't you link the others such as:

- Moment
- TinyColour
- React - PropTypes
and the others.

**Comment by BobbyWibowo** [comment]
I think all BD plugins suddenly stopped working today :thinking: 

Various errors in console, likely related to this PR:
```js
Uncaught Error: Could not fire onSwitch for plugin ServerFolders.
    at Function.fireEvent (/home/bobby/.config/EnhancedDiscord/bd_shit.js:320:35)
    at onSwitch (/home/bobby/.config/EnhancedDiscord/bd_shit.js:287:19)
    at CallbacksRegistry.apply (/usr/share/discord/resources/electron.asar/common/api/callbacks-registry.js:48:42)
    at EventEmitter.<anonymous> (/usr/share/discord/resources/electron.asar/renderer/api/remote.js:299:21)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)

Uncaught (in promise) TypeError: Cannot read property 'appendChild' of null
    at Function.showToast (/home/bobby/.config/EnhancedDiscord/dom_shit.js:297:43)
    at a._toast (/home/bobby/.config/EnhancedDiscord/plugins/magane.plugin.js:22:44285)
    at a.n [as _toast] (/home/bobby/.config/EnhancedDiscord/plugins/magane.plugin.js:22:105382)
    at a.<anonymous> (/home/bobby/.config/EnhancedDiscord/plugins/magane.plugin.js:22:50525)
    at g (https://discordapp.com/assets/47f13151b80452ea43a6.js:135:46817)
    at Generator._invoke (https://discordapp.com/assets/47f13151b80452ea43a6.js:135:46582)
    at Generator.e.(anonymous function) [as next] (https://discordapp.com/assets/47f13151b80452ea43a6.js:135:46996)
    at r (/home/bobby/.config/EnhancedDiscord/plugins/magane.plugin.js:22:42737)
    at /home/bobby/.config/EnhancedDiscord/plugins/magane.plugin.js:22:42831
```

The ones below are probably irrelevant, but they also broke down today:
```js
AutoCorrect fatal error, plugin could not be started! TypeError: Cannot read property 'nextSibling' of undefined
    at Object.tryCreateToastContainer (https://rawgit.com/Metalloriff/BetterDiscordPlugins/master/Lib/NeatoBurritoLibrary.js:2534:99)
    at Object.showToast (https://rawgit.com/Metalloriff/BetterDiscordPlugins/master/Lib/NeatoBurritoLibrary.js:2557:8)
    at Object.onPluginLoaded (https://rawgit.com/Metalloriff/BetterDiscordPlugins/master/Lib/NeatoBurritoLibrary.js:1761:13)
    at AutoCorrect.onLibLoaded (/home/bobby/.config/EnhancedDiscord/plugins/AutoCorrect.plugin.js:498:19)
    at HTMLScriptElement.libLoadedEvent (/home/bobby/.config/EnhancedDiscord/plugins/AutoCorrect.plugin.js:36:23)
    at HTMLScriptElement.o (https://discordapp.com/assets/47f13151b80452ea43a6.js:135:19839)

Uncaught TypeError: Cannot create proxy with a non-object as target or handler
    at new Proxy (<anonymous>)
    at mwittrien.github.io/BetterDiscordAddons/Plugins/BDFDB.js:1
    at mwittrien.github.io/BetterDiscordAddons/Plugins/BDFDB.js:1

Uncaught TypeError: BDFDB.getLibraryStrings is not a function
    at Object.BDFDB.loadMessage (https://mwittrien.github.io/BetterDiscordAddons/Plugins/BDFDB.js:1:975)
    at ShowImageDetails.initialize (/home/bobby/.config/EnhancedDiscord/plugins/ShowImageDetails.plugin.js:89:10)
    at HTMLScriptElement.libraryScript.addEventListener (/home/bobby/.config/EnhancedDiscord/plugins/ShowImageDetails.plugin.js:84:59)
    at HTMLScriptElement.o (https://discordapp.com/assets/47f13151b80452ea43a6.js:135:19839)

```

**Comment by rauenzi** [comment]
Now includes using `injection.js` from the EnhancedDiscord directory using `process.env.injDir`.

BD compatibility now only required when needed.

Style import switched back to website.

**Comment by rauenzi** [comment]
This PR fixes #56 and closes #54
